x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrailTests.java

```diff
@@ -279,6 +343,738 @@ public void testAuditFilterSettingValidation() {
+                .put(LoggingAuditTrail.REQUEST_ID_FIELD_NAME, requestId);
+        assertMsg(generatedDeleteRoleAuditEventString, checkedFields.immutableMap());
+    }
+
+    public void testSecurityConfigChangeEventFormattingForApiKeyInvalidation() throws IOException {
+        final String requestId = randomRequestId();
+        final String[] expectedRoles = randomArray(0, 4, String[]::new, () -> randomBoolean() ? null : randomAlphaOfLengthBetween(1, 4));
+        final AuthorizationInfo authorizationInfo = () -> Collections.singletonMap(PRINCIPAL_ROLES_FIELD_NAME, expectedRoles);
+        final Authentication authentication = createAuthentication();
+
+        final InvalidateApiKeyRequest invalidateApiKeyRequest;
+        if (randomBoolean()) {
+            invalidateApiKeyRequest = new InvalidateApiKeyRequest(randomFrom(randomAlphaOfLength(8), null),
+                    randomFrom(randomAlphaOfLength(8), null), null, randomFrom(randomAlphaOfLength(8), null), randomBoolean(),
+                    randomFrom(randomArray(3, String[]::new, () -> randomAlphaOfLength(8)), null));
+        } else {
+            invalidateApiKeyRequest = new InvalidateApiKeyRequest(randomFrom(randomAlphaOfLength(8), null),
+                    randomFrom(randomAlphaOfLength(8), null), randomAlphaOfLength(8), randomFrom(randomAlphaOfLength(8), null),
+                    randomBoolean());
+        }
+        auditTrail.accessGranted(requestId, authentication, InvalidateApiKeyAction.NAME, invalidateApiKeyRequest, authorizationInfo);
+        List<String> output = CapturingLogger.output(logger.getName(), Level.INFO);
+        assertThat(output.size(), is(2));
+        String generatedInvalidateKeyAuditEventString = output.get(1);
+        StringBuilder invalidateKeyEventStringBuilder = new StringBuilder();
+        invalidateKeyEventStringBuilder.append("\"invalidate\":{\"apikeys\":{");
+        if (invalidateApiKeyRequest.getIds() != null && invalidateApiKeyRequest.getIds().length > 0) {
+            invalidateKeyEventStringBuilder.append("\"ids\":[");
+            for (String apiKeyId : invalidateApiKeyRequest.getIds()) {
+                invalidateKeyEventStringBuilder.append("\"").append(apiKeyId).append("\",");
+            }
+            // delete last comma
+            invalidateKeyEventStringBuilder.deleteCharAt(invalidateKeyEventStringBuilder.length() - 1);
+            invalidateKeyEventStringBuilder.append("],");
+        }
+        if (Strings.hasLength(invalidateApiKeyRequest.getName())) {
+            invalidateKeyEventStringBuilder.append("\"name\":\"").append(invalidateApiKeyRequest.getName()).append("\",");
+        }
+        invalidateKeyEventStringBuilder.append("\"owned_by_authenticated_user\":")
+                .append(invalidateApiKeyRequest.ownedByAuthenticatedUser());
+        if (Strings.hasLength(invalidateApiKeyRequest.getUserName()) || Strings.hasLength(invalidateApiKeyRequest.getRealmName())) {
```

Comment:
Some of the usage of `Strings#hasLength` needs to be updated if it is changed in the main code.